python/pyspark/testing/utils.py

```diff
@@ -221,7 +221,130 @@ def check_error(
+                return False
+        return True
+
+    def compare_structfields_ignore_nullable(df_structfield, expected_structfield):
+        if ignore_nullable:
+            if df_structfield is None and expected_structfield is None:
+                return True
+            elif df_structfield is None or expected_structfield is None:
+                return False
+            if df_structfield.name != expected_structfield.name:
+                return False
+            else:
+                return compare_datatypes_ignore_nullable(
+                    df_structfield.dataType, expected_structfield.dataType
+                )
+        else:
+            return df_structfield == expected_structfield
+
+    def compare_datatypes_ignore_nullable(dt1, dt2):
+        # checks datatype equality, using recursion to ignore nullable
+        if dt1.typeName() == dt2.typeName():
+            if dt1.typeName() == "array":
+                return compare_datatypes_ignore_nullable(dt1.elementType, dt2.elementType)
+            elif dt1.typeName() == "struct":
+                return compare_schemas_ignore_nullable(dt1, dt2)
+            else:
+                return True
+        else:
+            return False
+
+    schemas_equal = True
+    error_msg = "Schemas do not match: \n"
+
+    if ignore_nullable:
+        if not compare_schemas_ignore_nullable(df_schema, expected_schema):
+            zipped = zip_longest(df_schema, expected_schema)
+            for df_structfield, expected_structfield in zipped:
+                if not compare_structfields_ignore_nullable(df_structfield, expected_structfield):
+                    schemas_equal = False
+                    error_msg += (
+                        "[df]"
+                        + "\n"
+                        + str(df_structfield)
+                        + "\n\n"
+                        + "[expected]"
+                        + "\n"
+                        + str(expected_structfield)
+                        + "\n\n"
+                    )
+    else:
+        if df_schema != expected_schema:
+            schemas_equal = False
+            zipped = zip_longest(df_schema, expected_schema)
+            for df_structfield, expected_structfield in zipped:
+                if df_structfield != expected_structfield:
+                    error_msg += (
+                        "[df]"
+                        + "\n"
+                        + str(df_structfield)
+                        + "\n\n"
+                        + "[expected]"
+                        + "\n"
+                        + str(expected_structfield)
+                        + "\n\n"
+                    )
+
+    if not schemas_equal:
+        raise PySparkAssertionError(
+            error_class="DIFFERENT_SCHEMA",
+            message_parameters={"error_msg": error_msg},
+        )
+
+
+def assertDataFrameEqual(
+    df: DataFrame, expected: DataFrame, check_row_order: bool = False, ignore_nullable: bool = False
```

Comment:
ignore_nullable -> check_schema_nullability? To make it consistent with `check_row_order`

Reply:
sounds good, how about `check_nullability` -- if set to False, it also will ignore nullable flag in Maps, Arrays, and Structs

Reply:
I'm wondering if the `check_nullability` option could be confusing for users, since it's unclear what the nullability is referring to. Is this option necessary?

Reply:
One option is to add a sth like `StructType.asNullable` that Scala side has, (and document that this is Unstable API). Then users can call

```python
assertDataFrameEqual(schema1.asNullable, schema2.asNullable)
```

Reply:
I am good with excluding nullability in this PR scope for now, but wonder if we can have a good error message for deeply nested type. That's actually my major concern.

Reply:
+1 on excluding the nullability check for this PR.